pkg/server/plugin/keymanager/awskms/awskms.go

```diff
@@ -0,0 +1,502 @@
+	res := keyEntry{}
+	description := p.descriptionFromSpireKeyID(spireKeyID)
+	keySpec, err := keySpecFromKeyType(keyType)
+	if err != nil {
+		return res, err
+	}
+
+	createKeyInput := &kms.CreateKeyInput{
+		Description:           aws.String(description),
+		KeyUsage:              types.KeyUsageTypeSignVerify,
+		CustomerMasterKeySpec: keySpec,
+	}
+
+	key, err := p.kmsClient.CreateKey(ctx, createKeyInput)
+	if err != nil {
+		return res, kmsErr.New("failed to create key: %v", err)
+	}
+
+	pub, err := p.kmsClient.GetPublicKey(ctx, &kms.GetPublicKeyInput{KeyId: key.KeyMetadata.KeyId})
+	if err != nil {
+		return res, kmsErr.New("failed to get public key: %v", err)
+	}
+
+	return keyEntry{
+		KMSKeyID: *pub.KeyId,
+		Alias:    p.aliasFromSpireKeyID(spireKeyID),
+		PublicKey: &keymanager.PublicKey{
+			Id:       spireKeyID,
+			Type:     keyType,
+			PkixData: pub.PublicKey,
+		},
+	}, nil
+}
+
+func (p *Plugin) buildKeyEntry(ctx context.Context, alias *string, awsKeyID *string) (*keyEntry, error) {
+	spireKeyID, ok := p.spireKeyIDFromAlias(*alias)
+	if !ok {
+		return nil, nil
+	}
+
+	describeResp, err := p.kmsClient.DescribeKey(ctx, &kms.DescribeKeyInput{KeyId: alias})
+	if err != nil {
+		return nil, kmsErr.New("failed to describe key: %v", err)
+	}
+
+	if !describeResp.KeyMetadata.Enabled {
+		return nil, nil
+	}
+
+	keyType, err := keyTypeFromKeySpec(describeResp.KeyMetadata.CustomerMasterKeySpec)
+	if err != nil {
+		p.log.Debug("Skipped key", "reason", err)
+		return nil, nil
+	}
+
+	getPublicKeyResp, err := p.kmsClient.GetPublicKey(ctx, &kms.GetPublicKeyInput{KeyId: alias})
+	if err != nil {
+		return nil, kmsErr.New("failed to get public key: %v", err)
+	}
+
+	return &keyEntry{
+		KMSKeyID: *awsKeyID,
+		Alias:    *alias,
+		PublicKey: &keymanager.PublicKey{
+			Id:       spireKeyID,
+			Type:     keyType,
+			PkixData: getPublicKeyResp.PublicKey,
+		},
+	}, err
+}
+
+func (p *Plugin) fetchAliasesPage(ctx context.Context, marker *string) (*string, error) {
+	aliasesResp, err := p.kmsClient.ListAliases(ctx, &kms.ListAliasesInput{
+		Marker: marker,
+	})
+	if err != nil {
+		return nil, kmsErr.New("failed to fetch keys: %v", err)
+	}
+
+	p.log.Debug("Found aliases", "num_aliases", len(aliasesResp.Aliases))
+
+	for _, alias := range aliasesResp.Aliases {
+		if alias.AliasName == nil || alias.TargetKeyId == nil {
+			continue
+		}
+
+		entry, err := p.buildKeyEntry(ctx, alias.AliasName, alias.TargetKeyId)
+		switch {
+		case err != nil:
+			return nil, kmsErr.New("failed to process KMS key: %v", err)
+		case entry != nil:
+			err := p.setEntry(entry.PublicKey.Id, *entry)
+			if err != nil {
+				return nil, err
+			}
+			p.log.Debug("Loaded key", keyIDTag, *alias.TargetKeyId, aliasTag, *alias.AliasName)
+		}
+	}
+	return aliasesResp.NextMarker, nil
+}
+
+func (p *Plugin) spireKeyIDFromAlias(alias string) (string, bool) {
+	prefix := aliasPrefix + p.keyPrefix
```

Comment:
If you mean `aliasPrefix` that is an implementation detail from AWS (at least that's how I understand it). I'm not comfortable about offering that as configurable. I think it's unlikely for AWS to change this and it might confuse users about why it's offered as a configuration.